### ISSUE TITLE
[Bug]: Wrong signal spacing while using autoplacement feature

### ISSUE THREAD

**Issue by mathan0305** [OpenTTD/OpenTTD#10450]
### Version of OpenTTD

1.13.0, Windows 11 22H2 x64

### Expected result

all signals have the same gap between them 

![image](https://user-images.githubusercontent.com/78717276/216853225-ecb716ef-65ec-437e-9e51-a669772123ba.png)

### Actual result

marked gap is larger than it should be. also works with any gap length. 

![image](https://user-images.githubusercontent.com/78717276/216853278-a1ee9db6-a0bd-4277-9172-4729d8bc873b.png)

### Steps to reproduce

1. turn on this setting.
![image](https://user-images.githubusercontent.com/78717276/216854044-c020aeff-1e03-47f5-9e3d-4ae35a2ab5b2.png)
2. build this rail pattern 
![image](https://user-images.githubusercontent.com/78717276/216854251-6203d6a2-7689-4990-a71c-4de8fb032919.png)
3. use autoplacing with distance = 2 from any side of this pattern 
4. result 
![image](https://user-images.githubusercontent.com/78717276/216854304-90590862-0a6d-4c2a-b75f-4603822a7e13.png)

also works on last nightly version of the game (openttd-20230205-master-g7c998d46bd-windows)

**Comment by ldpl** [comment]
I'd say it works as expected. Diagonal segments have different length so it's impossible to keep exactly the same gap. And you specifically enabled the setting that makes it more liberal with signal spacing. Combined with the fact that you're building signal backwards it leads to a situation where one diagonal piece combined with the straight one is less than your signal gap so signal is not built until one more segment is added. So if you want the game to make tighter gaps use the setting value that's designed to guarantee it.

**Comment by 2TallTyler** [comment]
@mathan0305 Can you please turn off the "keep fixed distance between signals" setting and report back? It should fix your problem. The setting is not well-named, but the helptext explains what it does.

**Comment by mathan0305** [comment]
@2TallTyler Yes, it helped. I read original english description and understood what is the problem. With russian translation that helptext means different thing. Something like that: "Without this setting signals are additionaly placed before and after tunnels, bridges. With it signals are placed every N-th tile" 
Just translation mistake

**Comment by ReluctantOwl** [comment]
Version: 13.0
Windows: 22H2 x64

I am getting a behaviour similar to this. Happy to raise as a separate request if needed, but this is still signal spacing related.

I usually an automatic placement of 8 which leaves a space of 7 tiles between signals. This worked fine in version 12.2 so something has been introduced in 13.0. In the image below the central "straight" line has a spacing between signals of 7 tiles. The two "diagonal" lines either side of it have the "keep fixed distance between signals" switched on and off as per @2TallTyler 's suggestion. As can be seen in the image the signals on the "diagonals" don't align with those on the "straight" line. I think they are 1 short, but not sure. 

![Example](https://user-images.githubusercontent.com/124810791/217608188-2f37bfe1-e50b-4c85-b071-df1952d6a7a1.png)

It happens with different auto-spacing too. I have also tried without GRFs and the same thing happens there.

**Comment by krysclarke** [comment]
> Version: 13.0 Windows: 22H2 x64
> 
> I am getting a behaviour similar to this. Happy to raise as a separate request if needed, but this is still signal spacing related.
> 
> I usually an automatic placement of 8 which leaves a space of 7 tiles between signals. This worked fine in version 12.2 so something has been introduced in 13.0. In the image below the central "straight" line has a spacing between signals of 7 tiles. The two "diagonal" lines either side of it have the "keep fixed distance between signals" switched on and off as per @2TallTyler 's suggestion. As can be seen in the image the signals on the "diagonals" don't align with those on the "straight" line. I think they are 1 short, but not sure.
> 
> ![Example](https://user-images.githubusercontent.com/124810791/217608188-2f37bfe1-e50b-4c85-b071-df1952d6a7a1.png)
> 
> It happens with different auto-spacing too. I have also tried without GRFs and the same thing happens there.

The spacing seems to be fine in that screenshot, from what I can see.

**Comment by ReluctantOwl** [comment]
> > Version: 13.0 Windows: 22H2 x64
> > I am getting a behaviour similar to this. Happy to raise as a separate request if needed, but this is still signal spacing related.
> > I usually an automatic placement of 8 which leaves a space of 7 tiles between signals. This worked fine in version 12.2 so something has been introduced in 13.0. In the image below the central "straight" line has a spacing between signals of 7 tiles. The two "diagonal" lines either side of it have the "keep fixed distance between signals" switched on and off as per @2TallTyler 's suggestion. As can be seen in the image the signals on the "diagonals" don't align with those on the "straight" line. I think they are 1 short, but not sure.
> > ![Example](https://user-images.githubusercontent.com/124810791/217608188-2f37bfe1-e50b-4c85-b071-df1952d6a7a1.png)
> > It happens with different auto-spacing too. I have also tried without GRFs and the same thing happens there.
> 
> The spacing seems to be fine in that screenshot, from what I can see.

Was on holiday so didn't see this response until now. Unfortunately whilst the spacing may initially look alright it isn't. The signals placed on the diagonals should line up with the signals on the centre line, unfortunately they do not.

**Comment by krysclarke** [comment]
> > > Version: 13.0 Windows: 22H2 x64
> > > I am getting a behaviour similar to this. Happy to raise as a separate request if needed, but this is still signal spacing related.
> > > I usually an automatic placement of 8 which leaves a space of 7 tiles between signals. This worked fine in version 12.2 so something has been introduced in 13.0. In the image below the central "straight" line has a spacing between signals of 7 tiles. The two "diagonal" lines either side of it have the "keep fixed distance between signals" switched on and off as per @2TallTyler 's suggestion. As can be seen in the image the signals on the "diagonals" don't align with those on the "straight" line. I think they are 1 short, but not sure.
> > > ![Example](https://user-images.githubusercontent.com/124810791/217608188-2f37bfe1-e50b-4c85-b071-df1952d6a7a1.png)
> > > It happens with different auto-spacing too. I have also tried without GRFs and the same thing happens there.
> > 
> > 
> > The spacing seems to be fine in that screenshot, from what I can see.
> 
> Was on holiday so didn't see this response until now. Unfortunately whilst the spacing may initially look alright it isn't. The signals placed on the diagonals should line up with the signals on the centre line, unfortunately they do not.

The calculation for diagonals does involve the square-root of 2, so an 8-tile spacing will place signals on every 6th grid-row & grid-column, like you see in your screenshot.

**Comment by lcd047** [comment]
There's a simple example of how the new behavior breaks game play.   Set signal dragging distance to 2, and take a train of length 3.  Up to 12.2 said train could fit between 3 consecutive signals regardless of whether the lines were built diagonally or not, and regardless of how many bends there were between those signals.  That's no longer the case with 13.0, a train on length 3 no longer fits between 3 consecutive signals on diagonals.  Building waiting bays and the like now involves moving signals manually, which is grinding at best.

True, length calculations for diagonals involve `sqrt(2)`, but the same is true for trains, it doesn't only apply to lines and signals.  Trains in OpenTTD are getting "longer" when running on diagonals.  This has been the case for ages, it's a problem that has been well understood from the very beginning, and no reasonable solution has been found for it so far.  If you want to fix this (for "realism", or whatever it was that motivated this change) then please do it also for the trains, not just for placing signals.  Or at least make it _optional_, as it is it breaks game play for many (perhaps most?) experienced players.

**Comment by ldpl** [comment]
@lcd047 Can you please add a screenshot to show what you mean? Anyway, this change was introduced in #9652 so considering prior to it you that those who care about exact spacing had to place signals manually on any diagonal doing it only in rare cases I'd count to be an improvement.

Btw, it doesn't involve sqrt(2) because nothing in this game is that simple xD Unless your sqrt(2) is 1.5 :p

**Comment by lcd047** [comment]
> @lcd047 Can you please add a screenshot to show what you mean?

![OpenTTD 12 2](https://user-images.githubusercontent.com/1281736/220058648-63b71acf-eac8-4260-a3bf-ac0d5ba0c6ce.png)
![OpenTTD 13 0](https://user-images.githubusercontent.com/1281736/220058663-59257023-57ab-41a7-abad-914e77b6a82e.png)

> Anyway, this change was introduced in #9652 so considering that prior to it "experienced players" had to place signals manually on any diagonal doing it only in rare cases I'd count to be an improvement.

Right: with trains with length 2.5, 3.5, 4.5, etc. you get the opposite problem.  So make it optional, to allow us to choose our poison. :smile:

> Btw, it doesn't involve sqrt(2)

It does in real life.  Then again, train stations in real life are seldom bigger than a city, so yeah, it isn't that simple.

**Comment by ldpl** [comment]
@lcd047 (and similarly @ReluctantOwl I guess). This doesn't explain your use case. Why does that matter that train fits between 3 signals instead of 4? What are you trying to achieve with it and how often is it really needed? I my 10+ years of playing OpenTTD I don't think I've ever seen anyone build diagonal waiting bays, for example.

It would be impossible for the game to move forward if it was piling up compatibility options for every change without considering how useful the old behaviour was. In this case, old signal placer didn't even follow any apparent logic, it could make the gap smaller or larger than requested even if counted by train length instead of speed. So even if the option is to be added it would be if it had a clear placement logic that fits as many players as possible. Can't exactly be adding an option for "ldc047 TL3 gameplay" :p There is already an awkward situation with "keep fixed distance" setting as it's unclear what exactly it was supposed to achieve but it's also hard to remove without knowing how many players are using it.

Currently it work like this:
- (fixed distance off). Signal placer is placing signals at a gap of N straight rail segments or less. If that's impossible (tunnel/bridge) it places them as close as it can. - (fixed distance on). Signal placer is trying to place signals at a regular interval of N straight rail pieces (rounded up to the nearest signal spot in case of diagonal) regardless of whether it's actually possible to bulid signal there.

Gap is calculated by the logical distance of the rail segment (as in movement calculation) in both cases.

**Comment by lcd047** [comment]
> @lcd047 (and similarly @ReluctantOwl I guess). This doesn't explain your use case. Why does that matter that train fits between 3 signals instead of 4?

Because waiting bays have to fit an entire number of trains.  With the old way, a bay with `2k+1` signals can fit `k` trains, regardless of how it bends.  With the new way a line with `2k+1` signals may or may not fit an entire number of trains, depending on how it bends.  And I now get to count manually where each train fits, which is great fun. :smile:

On a side note: the old way was consistent with the Manhattan distance that was used everywhere else in the game.  That's no longer the case now, which is essentially what the OP is saying.  For what it's worth.

> What are you trying to achieve with it

Size waiting bays so that the tail of the last train doesn't block the entry junction.

> and how often is it really needed?

The way I'm playing, pretty much all the time. _shrug_

> I my 10+ years of playing OpenTTD I don't think I've ever seen anyone build diagonal waiting bays, for example.

You do realize you can't make waiting bays with choices without at least _some_ diagonal tiles, right?

> It would be impossible for the game to move forward if it was piling up compatibility options for every change without considering how useful the old behaviour was.

Right: if you don't see the need for it then it _has_ to be discarded.  Always the best argument. :smile:

> In this case, old signal placer didn't even follow any apparent logic,

Manhattan distance.  Does that ring any bell?

> Currently it work like this:
> 
> * (fixed distance off). Signal placer is placing signals at a gap of N straight rail segments or less. If that's impossible (tunnel/bridge) it places them as close as it can (i.e. on bridge/tunnel ends).
> * (fixed distance on). Signal placer is trying to place signals at a regular interval of N straight rail pieces or more regardless of whether it's actually possible to bulid signal there.
> 
> Gap is calculated by the logical distance of the rail segment (as in train movement calculation) in both cases.

"Fixed distance" is about how signals sync after an obstacle (i.e. a tunnel or a bridge).  It has nothing to do with the discussion above.

**Comment by ldpl** [comment]
> > In this case, old signal placer didn't even follow any apparent logic,
> 
> Manhattan distance. Does that ring any bell?

Do you know what manhattan distance even is? :p Straight peice of rail has the same manhattan distance as a diagonal one yet it fits twice as many wagons. Besides, old placer wasn't even using any "distance" so it wasn't consistent no matter how you measure it.
![Screenshot from 2023-02-20 14-30-38](https://user-images.githubusercontent.com/413570/220080845-33a160c1-1c90-41a5-9d04-6a39f3fc9c52.png)

**Comment by lcd047** [comment]
> Do you know what manhattan distance even is?

I kind of do, yes.  The point is, it doesn't depend on the number of bends. :smile:

**Comment by ldpl** [comment]
> I kind of do, yes. The point is, it doesn't depend on the number of bends. smile
That's a meaningless point as no distance depends on the path by definition.

**Comment by lcd047** [comment]
> That's a meaningless point as no distance (aka metric) depends on the path by definition.

Right.  Since you insist on being pedantic: with Manhattan metrics, the distance traveled by a vehicle doesn't depend on the path taken.  That's not the case for other metrics: with the Euclidean metric f.i. the distance traveled by a vehicle between A and B depends very much on the path taken: the closer it is to a straight line, the shorter the distance.  That's why it's appropriate to use Manhattan in OpenTTD: because it's trivial to calculate.  And that's why it makes no sense to bring arguments related to Euclidean distance in this discussion: because Euclidean distance is a very different beast, that plays by very different rules. _shrug_

**Comment by ldpl** [comment]
@lcd047 you're confusing metric with the length of a path. No metric depends on the path, and two paths between two points can easily have different lengths in any metric.

Also, the game uses many metrics:
- Manhattan distance: `dx+dy`, used in profit calculations, kind of irrelevant for spacing.
- "Logical" distance: `min(dx, dy) * 2 + max(dx, dy) * 3` is used in train movement calculations.
- "Wagon" distance: `min(dx, dy)` is used for the train length and is what old placer can be argued to somewhat use. This one is what you seem to be confusing with manhattan.
- Euclidean metric: `sqrt(dx * dx + dy * dy)`. Used for town zones, totally unrelated to this topic.

What I see as a heart of this problem is that movement uses one metric while train length uses another. So when trains are moving one gap makes the most sense, when they're stopped - another.

**Comment by ReluctantOwl** [comment]
So for me, I usually have my signals at 8 spacing across my network, and I usually have a typical max train length of 7.

On the straights the way the distance between signals is measured when using the drag method is fine no issues, the numbers work.

however, as soon as I get to the diagonals the signals get pushed together because at some point between 12.2 and 13.0 the way that the distance on diagonals changed. This is all regardless of minimising the gaps, i'm talking open track, no tunnels, bridges or junctions. The result of this as a direct impact of the change is that a train of length 7 now sits over 2 signal blocks rather than fitting into one when on diagonals. This is alright when there is no traffic but becomes wasteful when there is.

To me the logical expectation when using the auto-placement feature would be that the physical distance between two signals, or signal block length, remains constant. Yet we now have a system where when using 8 (Using 8 because that is what I know) on a straight there are 7 empty spaces between 2 signals but on a diagonal we get 11 half space, equivalent to 5.5 empty spaces between the two signals. I don't think any amount of explaining the varying ways to calculate distance makes this make sense. Auto-placer is meant to make things quick and easy, then suddenly you have to know two different rules based on the diagonals.

That said, I have only today found out about ctrl+drag of signals to do the whole line, I have used just drag until this point, which still leads to poor spacing!

Hope this helps *shrug*

**Comment by ldpl** [comment]
> The result of this as a direct impact of the change is that a train of length 7 now sits over 2 signal blocks rather than fitting into one when on diagonals. This is alright when there is no traffic but becomes wasteful when there is.

This makes no sense, train always takes at least two signal blocks when it moves. In fact, this change made the gap smaller on diagonal, so now trains can follow each other more closely than before. If check the PR it all started because before you had to add signals manually to all diagonals to make trains not stop and jam when turning.

> To me the logical expectation when using the auto-placement feature would be that the physical distance between two signals, or signal block length, remains constant. Yet we now have a system where when using 8 (Using 8 because that is what I know) on a straight there are 7 empty spaces between 2 signals but on a diagonal we get 11 half space, equivalent to 5.5 empty spaces between the two signals. I don't think any amount of explaining the varying ways to calculate distance makes this make sense. Auto-placer is meant to make things quick and easy, then suddenly you have to know two different rules based on the diagonals.

There is no quick an easy. For historical reasons game doesn't operate with distances like you would expect in a normal (euclidean) space. So, in a way, trains become longer when they turn on diagonal and shorter when they turn back. So if you measure signal distance by the way train looks it won't be the same as if you measure how the train moves. This change just picked function over form, it's more important to make sure that signals are placed it way they don't stop trains needlessly than to make them equally spaced when compared to train length.

**Comment by ReluctantOwl** [comment]
> 

> This makes no sense, train always takes at least two signal blocks when it moves. In fact, this change made the gap smaller on diagonal, so now trains can follow each other more closely than before. If check the PR it all started because before you had to add signals manually to all diagonals to make trains not stop and jam when turning.

My bad I did mean a stationary train, a train held at a signal used to fit between two signals that were auto-spaced now because the auto-space has some new way of measuring diagonals a held train crosses over 2 signal blocks. This does not occur on a straight.

I also think that because I have used a spacing of 8 as my default for somewhere around 10 years or so (who really knows how long it's been, I've been playing this game for at least 15 years), I have negated the issues of trains jamming on corners because a block length of 8 is quite large, the trains create enough separation that it doesn't become much of an issue.

> There is no quick and easy. For historical reasons game doesn't operate with distances like you would expect in a normal (euclidean) space. So, in a way, trains become longer when they turn on diagonal and shorter when they turn back. So if you measure signal distance by the way train looks it won't be the same as if you measure how the train moves. This change just picked function over form, it's more important to make sure that signals are placed it way they don't stop trains needlessly than to make them look equally spaced when compared to train length.

Right so having read back over the various posts of how distance is and isn't measured I think I have finally got my head round what it is you are trying to convey.

Which, I believe, is; The new way the game places the signals means that if it takes 10 seconds for a train travelling at full speed to go between signals on the straight it now takes 10 seconds on the diagonals before. 

Lets say that that distance happens to be 5 squares on the straight. It takes the train 10 seconds to travel over 5 squares. Not end to end of the train just the nose from end to end of the squares. 

In the previous versions the spacing on the diagonals would've been about 7 squares because of Pythagoras, (on the assumption that it is all diagonal) which means it takes a second or so longer to cross the diagonal distance.

What's been done on the new version is rather than keeping the spacing between the signals such that they are the same number of tiles apart, the calculation has been skewed such that a train travelling through the signals passes a signal every 10 seconds. i.e. they are physically 5 squares apart.

I see the logic, but damn that's kinda confusing when for years Trains fit between signals!

I concede that it probably isn't a bug, however, this really could've been explained better. Not sure exactly how but it isn't super clear that this is the new behaviour. I guess I will have to learn the new ways

> UPD: here is a good video illustrating this behaviour: https://cdn.discordapp.com/attachments/337701432230805505/901912563787386961/2021-10-24_22-14-38.mp4

Well that is just weird! I can see how that happens, that is probably something that should be fixed more than signal placement. Even changing where the signals are isn't going to help!

**Comment by Mons1990** [comment]
I've read both your POVs very with interest. However I completely agree with @lcd047. I hope this behavior will be rollbacked to 12.x as soon as possible. Same section lenght must result in same train occupancy, that's not metric or game logic, that's primary school physics. If trains were "compressed and expanded" in diagonal routes this change would have made sense but now - I'm sorry to say because I pay maximum respect for developers - it hasn't.

**Comment by ldpl** [comment]
>If trains were "compressed and expanded" in diagonal routes this change would have made sense 
But trains are compressed and expanded on diagonals. If they weren't they would've fit new (equal) distance.

**Comment by ldpl** [comment]
@Mons1990 I'm not confusing anything, watch the video I linked a few posts above. Trains are physically getting longer, no matter the projection. So there are 2 ways to place the signal one that is consistent with train movement and another that is consistent with their length. As this is supposedly a game about moving cargo use case for the former one is pretty apparent and is explained in the initial issue. And despite all the arguing there is a distinct lack of use cases for the latter one, so far I've only seen diagonal waiting bays being mentioned, but that's pretty niche. If more come up we can discuss possible solutions. But "school physics" doesn't make a use case, sorry.

**Comment by lcd047** [comment]
> And despite all the arguing there is a distinct lack of use cases for the latter one, so far I've only seen diagonal waiting bays being mentioned, but that's pretty niche.

It's so niche that it's relatively hard to even come up with a design of useful waiting bays that don't use _any_ diagonal tiles.  But, of course, if you guys are dead set to not see that then you won't see it, regardless of how much the rest of us point it out.  Soon you'll have telemetry as a more elegant tool to shut up any dissenting opinions about play style, so I suppose all this is a pointless struggle anyway. :+1:

**Comment by lcd047** [comment]
> Fun fact: @ldpl is not a dev and has no more influence on the way the game is changed (and certainly to "shut up any dissenting opinions") than anyone else.

Well, this part could (err, actually did, I suppose) have fooled me:

> If more come up we can discuss possible solutions. But "school physics" doesn't make a use case, sorry.

:smile:

**Comment by LordAro** [comment]
@ldpl is more often than not on "your side", i.e. with an opposing position to "the devs" (which, I would like to remind you, are by no means a cohesive group) than not. Please try to remember that people are people with their own set of opinions about everything and aren't exclusively united in one front against you and what you want.

I realise this is further derailing the issue at this point, but "telemetry" is being added explicitly so that we ("the devs") *can* be informed about how people play the game. I highly encourage you to read the details of what's being added in 14.x - TB wrote about it here https://www.openttd.org/news/2023/05/14/policy-and-survey , and the links contained within. Of course statistics can always be abused to point at whatever viewpoint you want, but our hope is that we can at least point to something concrete rather than all the useless "I think this", "I don't know anyone that uses that" annecdotal stuff that people usually spout about "playstyles".

**Comment by ldpl** [comment]
@lcd047 So far you're they only one with a valid use case here. Others just don't seem to understand why this change was made in the first place. And while "I just like the old way" is also an ok opinion to have it doesn't give any material for thought. So far, ignoring the nonconstructive "rollback" and "do nothing" options, I see at least 3 possible solutions:
1. Add a separate setting.
2. Extend "keep fixed distance" setting to cover waiting bay usecase.
3. Add a toggle to signal ui to switch between "moving" and "waiting" modes.
They all have their upsides and downsides and despite the lengthy discussion here I still don't see enough information to pick the best one. I guess main 2 questions to answer are:
1. Is this something that needs to be changed often during the game or is it just a personal preference for the entire game? "Waiting bay" usage suggests the former but then you're just suggesting to add the setting.
2. What is the use case for "keep fixed distance" setting? Is it related to "waiting bay" mode or is it totally independent and all 4 combinations make sense?

**Comment by Mons1990** [comment]
@ldpl I don't think you are in charge of decide who has valid use case and who hasn't, who understand and who doesn't and what is constructive and what isn't. The fact is that two people have an opinion and all the others have another. Thus I keep hoping in a rollback, now I stopped playing this game in favor of other ones because this is too much unconfortable for me. If there will be the possibility to choose which game the user prefers I will restore the 2.x settings and the discussion is closed. However I think OpenTTD has already a lot of advanced settings (in my opinion a bit too many) and before adding another one players for and against should be counted and considered. Finally I'd like to say that we're a community of players who aim at the same target: have fun; I think discussions like this shouldn't be taken so personally.

**Comment by Mons1990** [comment]
Has someone already taken in charge this issue?

**Comment by BackThePortal** [comment]
I've read all of this and I don't understand how this is so controversial, it seems to me that it's quite simple.

The gap length that players set in the Signals window should be in the same units as the number that shows up in the Depot window, and should work in all track directions.

It is my understanding that, in the depot, 1, that is, two standard wagons, equals to one diagonal track and 0.5, that is, one standard wagon, equals to one horizontal or vertical track. In my experience playing the game, this seems to be correct. We should then make the signals gap work the same way.

I'm really surprised by the time that has been wasted talking about this... Unless I'm the one who's completely wrong. Am I?

**Comment by BackThePortal** [comment]
There's also the length that appears when placing tracks or signals and dragging the mouse. The gap length should be the same as this number too. Also, this one is rounded up--I think it shouldn't be? Maybe I'll create a separate issue for this.
